File: frontend/types/github.ts
export interface GitHubRepository {
  id: number
  uuid: string
  name: string
  owner: string
  full_name: string
  description: string
  url: string
  is_private: boolean
  default_branch: string
  last_ingested_at: string | null
  ingestion_status: 'pending' | 'running' | 'completed' | 'failed'
  created_at: string
  updated_at: string
}

export interface GitHubIssue {
  id: number
  uuid: string
  github_id: number
  number: number
  title: string
  body: string
  state: 'open' | 'closed'
  author: string
  author_association: string
  assignees: string[]
  labels: string[]
  milestone: any
  locked: boolean
  created_at: string
  updated_at: string
  closed_at: string | null
  url: string
  comments: GitHubIssueComment[]
  comment_count: number
}

export interface GitHubIssueComment {
  id: number
  uuid: string
  github_id: number
  body: string
  author: string
  author_association: string
  created_at: string
  updated_at: string
  url: string
}

export interface GitHubPullRequest {
  id: number
  uuid: string
  github_id: number
  number: number
  title: string
  body: string
  state: 'open' | 'closed' | 'merged'
  author: string
  author_association: string
  assignees: string[]
  reviewers: string[]
  labels: string[]
  milestone: any
  head_branch: string
  base_branch: string
  merged: boolean
  merged_at: string | null
  merge_commit_sha: string
  additions: number
  deletions: number
  changed_files: number
  created_at: string
  updated_at: string
  closed_at: string | null
  url: string
  comments: GitHubPRComment[]
  files: GitHubPRFile[]
  comment_count: number
  file_count: number
}

export interface GitHubPRComment {
  id: number
  uuid: string
  github_id: number
  body: string
  author: string
  author_association: string
  created_at: string
  updated_at: string
  url: string
}

export interface GitHubPRFile {
  id: number
  uuid: string
  filename: string
  status: 'added' | 'modified' | 'removed' | 'renamed'
  additions: number
  deletions: number
  changes: number
  patch: string
  blob_url: string
  raw_url: string
  contents_url: string
}

export interface GitHubDiscussion {
  id: number
  uuid: string
  github_id: number
  number: number
  title: string
  body: string
  author: string
  author_association: string
  category: {
    id: number
    name: string
    emoji: string
    description: string
  }
  answer_chosen_at: string | null
  answer_chosen_by: string
  upvote_count: number
  viewer_has_upvoted: boolean
  created_at: string
  updated_at: string
  last_edited_at: string | null
  url: string
  comments: GitHubDiscussionComment[]
  comment_count: number
}

export interface GitHubDiscussionComment {
  id: number
  uuid: string
  github_id: number
  body: string
  author: string
  author_association: string
  created_at: string
  updated_at: string
  last_edited_at: string | null
  upvote_count: number
  viewer_has_upvoted: boolean
  parent_comment: number | null
  replies: GitHubDiscussionComment[]
}

export interface GitHubWikiPage {
  id: number
  uuid: string
  title: string
  content: string
  sha: string
  html_url: string
  download_url: string
  last_modified: string
}

export interface GitHubRepositoryFile {
  id: number
  uuid: string
  path: string
  name: string
  content: string
  sha: string
  size: number
  content_type: string
  encoding: string
  html_url: string
  download_url: string
  last_modified: string
}

export interface GitHubIngestionRequest {
  owner: string
  repo: string
  since?: string
  app_integration_id: number
}

export interface GitHubRepositoryResponse extends GitHubRepository {
  issue_count: number
  pr_count: number
  discussion_count: number
  wiki_page_count: number
  file_count: number
}

export interface GitHubIssuesResponse {
  count: number
  next: string | null
  previous: string | null
  results: GitHubIssue[]
}

export interface GitHubPullRequestsResponse {
  count: number
  next: string | null
  previous: string | null
  results: GitHubPullRequest[]
}

export interface GitHubIngestionStatus {
  status: 'success' | 'error' | 'pending'
  repository: string
  message: string
  completed_at?: string
  error?: string
}

export interface GitHubIngestionState {
  repositories: GitHubRepository[]
  currentRepository: GitHubRepository | null
  issues: GitHubIssue[]
  pullRequests: GitHubPullRequest[]
  loading: boolean
  error: string | null
  ingestionStatus: 'idle' | 'ingesting' | 'completed' | 'failed'
}

export interface GitHubFilters {
  state: 'all' | 'open' | 'closed' | 'merged'
  author: string
  labels: string[]
  since: string
  until: string
  search: string
}

export interface GitHubStats {
  totalIssues: number
  openIssues: number
  closedIssues: number
  totalPRs: number
  openPRs: number
  mergedPRs: number
  totalDiscussions: number
  totalWikiPages: number
  totalFiles: number
}

File: backend/core/services/providers/version_control/github.py
import logging
from typing import Dict, List, Optional, Any, Tuple

from core.services.providers.version_control import VCProvider, register_provider
from core.services.github_client import GitHubAPIClient

logger = logging.getLogger(__name__)


@register_provider('github')
class GitHubProvider(VCProvider):

    provider_name = 'github'

    def __init__(self, credentials: Dict[str, Any]):
        super().__init__(credentials)
        self._client: Optional[GitHubAPIClient] = None

    def _get_client(self) -> GitHubAPIClient:
        if not self._client:
            token = self.credentials.get('token')
            if not token:
                raise ValueError("GitHub token not found in credentials")
            self._client = GitHubAPIClient(token)
        return self._client

    def validate_credentials(self) -> Tuple[bool, str, Dict[str, Any]]:
        try:
            client = self._get_client()
            user_info = client._make_request('GET', '/user')
            metadata = {
                'login': user_info.get('login'),
                'name': user_info.get('name'),
                'avatar_url': user_info.get('avatar_url'),
                'html_url': user_info.get('html_url'),
            }
            return True, '', metadata
        except Exception as e:
            return False, str(e), {}

    def get_repository_info(self, owner: str, repo: str) -> Dict[str, Any]:
        client = self._get_client()
        data = client.get_repository_info(owner, repo)
        return {
            'id': str(data.get('id')),
            'name': data.get('name'),
            'owner': data.get('owner', {}).get('login'),
            'full_name': data.get('full_name'),
            'description': data.get('description', ''),
            'url': data.get('html_url'),
            'is_private': data.get('private', False),
            'default_branch': data.get('default_branch', 'main'),
            'fork': data.get('fork', False),
            'stars': data.get('stargazers_count', 0),
            'watchers': data.get('watchers_count', 0),
            'forks': data.get('forks_count', 0),
            'open_issues': data.get('open_issues_count', 0),
        }

    def get_issues(self, owner: str, repo: str, state: str = 'all',
                   since: Optional[str] = None) -> List[Dict[str, Any]]:
        client = self._get_client()
        issues = client.get_issues(owner, repo, state=state, since=since)
        return [
            {
                'id': str(issue['id']),
                'number': issue['number'],
                'title': issue['title'],
                'body': issue.get('body', '') or '',
                'state': issue['state'],
                'author': issue.get('user', {}).get('login', ''),
                'author_association': issue.get('author_association', ''),
                'assignees': [user['login'] for user in issue.get('assignees', [])],
                'labels': [label['name'] for label in issue.get('labels', [])],
                'milestone': issue.get('milestone'),
                'locked': issue.get('locked', False),
                'created_at': issue['created_at'],
                'updated_at': issue['updated_at'],
                'closed_at': issue.get('closed_at'),
                'url': issue['html_url'],
            }
            for issue in issues
        ]

    def get_issue_comments(self, owner: str, repo: str, issue_number: int) -> List[Dict[str, Any]]:
        client = self._get_client()
        comments = client.get_issue_comments(owner, repo, issue_number)
        return [
            {
                'id': str(comment['id']),
                'body': comment['body'],
                'author': comment.get('user', {}).get('login', ''),
                'author_association': comment.get('author_association', ''),
                'created_at': comment['created_at'],
                'updated_at': comment['updated_at'],
                'url': comment['html_url'],
            }
            for comment in comments
        ]

    def get_pull_requests(self, owner: str, repo: str, state: str = 'all',
                          since: Optional[str] = None) -> List[Dict[str, Any]]:
        client = self._get_client()
        prs = client.get_pull_requests(owner, repo, state=state, since=since)
        return [
            {
                'id': str(pr['id']),
                'number': pr['number'],
                'title': pr['title'],
                'body': pr.get('body', '') or '',
                'state': pr['state'],
                'author': pr.get('user', {}).get('login', ''),
                'author_association': pr.get('author_association', ''),
                'assignees': [user['login'] for user in pr.get('assignees', [])],
                'reviewers': [user['login'] for user in pr.get('requested_reviewers', [])],
                'labels': [label['name'] for label in pr.get('labels', [])],
                'milestone': pr.get('milestone'),
                'head_branch': pr.get('head', {}).get('ref', ''),
                'base_branch': pr.get('base', {}).get('ref', ''),
                'merged': pr.get('merged', False),
                'merged_at': pr.get('merged_at'),
                'merge_commit_sha': pr.get('merge_commit_sha', ''),
                'additions': pr.get('additions', 0),
                'deletions': pr.get('deletions', 0),
                'changed_files': pr.get('changed_files', 0),
                'created_at': pr['created_at'],
                'updated_at': pr['updated_at'],
                'closed_at': pr.get('closed_at'),
                'url': pr['html_url'],
            }
            for pr in prs
        ]

    def get_pull_request_comments(self, owner: str, repo: str, pr_number: int) -> List[Dict[str, Any]]:
        client = self._get_client()
        comments = client.get_pull_request_comments(owner, repo, pr_number)
        return [
            {
                'id': str(comment['id']),
                'body': comment['body'],
                'author': comment.get('user', {}).get('login', ''),
                'author_association': comment.get('author_association', ''),
                'created_at': comment['created_at'],
                'updated_at': comment['updated_at'],
                'url': comment['html_url'],
            }
            for comment in comments
        ]

    def get_pull_request_files(self, owner: str, repo: str, pr_number: int) -> List[Dict[str, Any]]:
        client = self._get_client()
        files = client.get_pull_request_files(owner, repo, pr_number)
        return [
            {
                'filename': file['filename'],
                'status': file['status'],
                'additions': file.get('additions', 0),
                'deletions': file.get('deletions', 0),
                'changes': file.get('changes', 0),
                'patch': file.get('patch', ''),
                'blob_url': file.get('blob_url', ''),
                'raw_url': file.get('raw_url', ''),
            }
            for file in files
        ]

    def close(self):
        if self._client:
            self._client.close()
            self._client = None
